packages/vite/src/node/server/pluginContainer.ts

```diff
@@ -576,13 +577,28 @@ class PluginContext implements Omit<RollupPluginContext, 'cache'> {
       skipSelf?: boolean
     },
   ) {
-    const skipCalls =
-      options?.skipSelf === false
-        ? this._resolveSkipCalls
-        : [
-            ...(this._resolveSkipCalls || []),
-            { id, importer, plugin: this._plugin },
-          ]
+    let skipCalls: readonly SkipInformation[] | undefined
+    if (options?.skipSelf === false) {
+      skipCalls = this._resolveSkipCalls
+    } else if (this._resolveSkipCalls) {
+      const skipCallsTemp = [...this._resolveSkipCalls]
+      const sameCallIndex = this._resolveSkipCalls.findIndex(
+        (c) =>
+          c.id === id && c.importer === importer && c.plugin === this._plugin,
+      )
+      if (sameCallIndex !== -1) {
+        skipCallsTemp[sameCallIndex] = {
+          ...skipCallsTemp[sameCallIndex],
+          called: true,
+        }
```

Comment:
Awesome investigation!
The fix looks good to merge, but while trying to understand the logic, I cannot tell why we cannot immediately do `return null` for such call, which seem to be destined to resolve loop.
The new test case seems to pass with "just `return null`". Do you have a case such approach fails?

Reply:
I added a test case that fails with `return null`: 67c4a661c36d758761d3c64c09b5a54247754d11